src/ssl/support.cc

```diff
@@ -213,14 +289,24 @@ int Ssl::matchX509CommonNames(X509 *peer_cert, void *check_data, int (*check_fun
         int numalts = sk_GENERAL_NAME_num(altnames);
         for (int i = 0; i < numalts; ++i) {
             const GENERAL_NAME *check = sk_GENERAL_NAME_value(altnames, i);
-            if (check->type != GEN_DNS) {
-                continue;
-            }
-            ASN1_STRING *cn_data = check->d.dNSName;
-
-            if ( (*check_func)(check_data, cn_data) == 0) {
-                sk_GENERAL_NAME_pop_free(altnames, GENERAL_NAME_free);
-                return 1;
+            switch(check->type) {
+                case GEN_DNS:
+                    // If the type is GEN_DNS, call check_func with the dNSName data
+                    if ( (*check_func)(check_data, check->d.dNSName) == 0) {
+                        sk_GENERAL_NAME_pop_free(altnames, GENERAL_NAME_free);
+                        return 1;
+                    }
+                    break;
+                case GEN_IPADD:
+                    debugs(83, 4, "Check type is GEN_IPADD, verifying...");
+                    // If it's an IP address, call check_func with the iPAddress data
+                    if ( (*check_func)(check_data, check->d.iPAddress) == 0) {
```

Comment:
N.B. The type of the second check_func() argument is (a pointer to) an ASN1_STRING. "Almost all ASN1 types in OpenSSL are represented as an ASN1_STRING structure. ASN1_STRING is also used for some CHOICE types." Roughly speaking, ASN1_STRING can store nearly anything.

I am concerned that this code loses "this ASN1_STRING is a GEN_IPADD SAN (rather than, say, GEN_DNS SAN)" knowledge and, hence, forces check_func() to _guess_ what information its second parameter actually contains. That guess is currently based on ASN1_STRING_type(asn1_data) not being V_ASN1_OCTET_STRING. I assume that other CN and SAN types may pass that rather weak "type check", in current or future code, resulting in silently incorrect casting and incorrect matching in check_func() calls that are not as well protected as this one.

I do understand that you would prefer not to fiddle with other check_func callers. However, this information loss and associated dangerous guesses is a very large red flag for me. If I am missing some built-in protections that make this loss and guesses safe, please do let me know. If this concern is not addressed (one way or the other), I will find the time to start addressing it myself after the next iteration, after you address other change requests (that do not require changing other check_func() callers).

Reply:
So in the latest commit, I have indeed fiddled with all of the callers. In all call sites, an explicit enum `Ssl::AddressType` is used to tell the `check_domain` function which address type it is using to remove the guess work. I hope this address the red flag concern.